plugins/ptp_operator/metrics/metrics.go

```diff
@@ -0,0 +1,381 @@
+			Subsystem: ptpSubsystem,
+			Name:      "max_offset_from_master",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+
+	// FrequencyAdjustment metrics to show frequency adjustment
+	FrequencyAdjustment = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "frequency_adjustment",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+
+	// DelayFromMaster metrics to show delay from the master
+	DelayFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "delay_from_master",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+)
+
+// Stats calculates stats  nolint:unused
+type Stats struct {
+	num                 float64
+	max                 float64
+	min                 float64
+	mean                float64
+	sumSqr              float64
+	sumDiffSqr          float64
+	frequencyAdjustment float64
+	delayFromMaster     float64
+	lastOffset          float64 //nolint:structcheck
+	lastClockState      ceevent.SyncState
+}
+
+func (s *Stats) addValue(val float64) {
+
+	oldMean := s.mean
+
+	if s.num == 0 || s.max < val {
+		s.max = val
+	}
+	if s.num == 0 && s.min > val {
+		s.mean = val
+	}
+	s.num++
+	s.mean = oldMean + (val-oldMean)/s.num
+	s.sumSqr += val * val
+	s.sumDiffSqr += (val - oldMean) * (val - s.mean)
+
+}
+
+// get stdDev
+func (s *Stats) getStdev() float64 { //nolint:unused
+	if s.num > 0 {
+		return math.Sqrt(s.sumDiffSqr / s.num)
+	}
+	return 1
+}
+func (s *Stats) getMaxAbs() float64 {
+	if s.max > s.mean {
+		return s.max
+	}
+	return s.min
+
+}
+
+// reset status
+func (s *Stats) reset() { //nolint:unused
+	s.num = 0
+	s.max = 0
+	s.mean = 0
+	s.sumDiffSqr = 0
+	s.sumSqr = 0
+}
+
+// Metric for PTP
+type Metric struct {
+	lock  sync.RWMutex
+	Stats map[string]*Stats
+}
+
+// ExtractMetrics ...
+func (m *Metric) ExtractMetrics(msg string) {
+	replacer := strings.NewReplacer("[", " ", "]", " ", ":", " ")
+	output := replacer.Replace(msg)
+	fields := strings.Fields(output)
+	if len(fields) < 2 {
+		glog.Errorf("ignoring log:log  is not in required format processname[time]: [iface]")
+		return
+	}
+	processName := fields[0]
+	iface := fields[2]
+	if strings.Contains(output, " max ") { // this get generated in case -u is passed an option to phy2sys opts
+		offsetFromMaster, maxOffsetFromMaster, frequencyAdjustment, delayFromMaster := extractSummaryMetrics(processName, output)
+		UpdatePTPMetrics(processName, iface, offsetFromMaster, maxOffsetFromMaster, frequencyAdjustment, delayFromMaster)
+	} else if strings.Contains(output, " offset ") { // dispatcher calls metrics for phy2sys
+		offsetFromMaster, maxOffsetFromMaster, frequencyAdjustment, delayFromMaster, clockState := extractRegularMetrics(processName, output)
+		// update stats
+		if clockState == "" {
+			return
+		}
+		var s *Stats
+		var found bool
+		if s, found = m.Stats[iface]; !found {
```

Comment:
Is a lock needed?  Is there another other thread that is modifying the the Stats map?

Reply:
Fixed added new state

Reply:
No, Also I need to remove the grouting call in the socket read function.
```
for {
		fd, err := l.Accept()
		if err != nil {
			log.Errorf("accept error: %s", err)
		} else {
			go processMessages(fd)
		}
	}
```